algebird-core/src/main/scala/com/twitter/algebird/CuckooFilter.scala

```diff
@@ -0,0 +1,435 @@
+/*
+Copyright 2012 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+ */
+
+package com.twitter.algebird
+
+import algebra.BoundedSemilattice
+import com.googlecode.javaewah.{IntIterator, EWAHCompressedBitmap => CBitSet}
+
+import scala.util.Random
+
+/**
+ * Cuckoo filter for Algebird in a Monoid way of life.
+ * referring :
+ *  - https://github.com/twitter/algebird/issues/560#issue-185161455
+ *  - http://www.eecs.harvard.edu/%7Emichaelm/postscripts/cuckoo-conext2014.pdf
+ *  - https://github.com/irfansharif/cfilter
+ * By nature, this filter isn't commutative
+ * From the inital paper, there is no problem to consider || fingerprint|| = ln(N) where N = fingerprintPerBucket * totalBucket
+ * "" as long as we use reasonably sized buckets, the fingerprint size can remain small. "", we'll use a a 32 bits fingerprint.
+ *
+ * We choose to put a fixed size for the fingerprint.
+ *
+ * TODO : Make cuckoo filter works like an aggregator like the BloomFilter.
+ * TODO : Lookup method have to return a Approximate number like the size method (sometimes you can't insert an element).
+  **/
+object CuckooFilter {
+  def apply[A](fingerprintPerBucket: Int, buckets: Int = 256)(
+      implicit hash1: Hash128[A]): CuckooFilterMonoid[A] = CuckooFilterMonoid(fingerprintPerBucket, buckets)
+}
+
+/**
+ * The cuckoo filter monoid
+  **/
+// TODO : Add the fingerprint size.
+case class CuckooFilterMonoid[A](fingerprintBucket: Int, totalBuckets: Int = 256)(implicit h1: Hash128[A])
+    extends Monoid[CF[A]]
+    with BoundedSemilattice[CF[A]] {
+
+  val cFHash: CFHash[A] = CFHash[A](totalBuckets)
+  val zero: CF[A] = CFZero(cFHash, fingerprintBucket, totalBuckets)
+
+  def plus(left: CF[A], right: CF[A]): CF[A] = left ++ right
+
+  /**
+   * it's come from the BloomFilter. But I don't think it's a good idea because insertion from scratch
+   * describe by the "set" and "add" functions are not safe. */
+  override def sumOption(iter: TraversableOnce[CF[A]]): Option[CF[A]] =
+    if (iter.isEmpty) None
+    else {
+      val buckets = Array.fill[CBitSet](totalBuckets)(new CBitSet(fingerprintBucket))
```

Comment:
It's more efficient to use 1 bitset of `numBuckets * numEntriesPerBucket * numBitsPerFingerprint`? IIRC JVM pointers are 8 bytes, so there's `totalBuckets * 8` overhead here, plus those from `CBitSet` and memory lookups.

According to the paper, each bucket should have `numEntriesPerBucket` slots, each with `numBitsPerFingerprint`. how does `fingerprintBucket` fit in here?